### PR TITLE
1. Added support for single level metrics aggregation (without groupby field) 2. Added 2 new methods and deprecated 6 methods

### PR DESCRIPTION
**Pull request by RenukaGurumurthy**

### CODE REVIEW

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java [hunk 1]
```diff
@@ -402,7 +409,14 @@ private void buildBuckets(String traceId,String index, RequestParamsDTO requestP
 		try {
 			TermsBuilder termBuilder = null;
 			String[] groupBy = requestParamsDTO.getGroupBy().split(APIConstants.COMMA);
-
+			FilterAggregationBuilder filterAggregationBuilder = null;
+			
+			if (getBaseAPIService().checkNull(requestParamsDTO.getFilter())) {
```
Comment (kumaraswamyS) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java]: @RenukaGurumurthy  Remove duplicate if condition.

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java [hunk 2]
```diff
@@ -414,21 +428,18 @@ private void buildBuckets(String traceId,String index, RequestParamsDTO requestP
...
 				}
 			}
-			if (getBaseAPIService().checkNull(requestParamsDTO.getFilter())) {
-				FilterAggregationBuilder filterBuilder = null;
-				if (filterBuilder == null) {
-					filterBuilder = includeFilterAggregate(index, requestParamsDTO.getFilter(),validatedData, filterData, userFilter);
-				}
+				
+			if(filterAggregationBuilder  != null) {
 				if (termBuilder != null) {
```
Comment (kumaraswamyS) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java]: @RenukaGurumurthy if and else both condition will create the termBuilder object, so remove the not null statement.

File: src/main/java/org/gooru/insights/services/ESDataProcessorServiceImpl.java [hunk 3]
```diff
@@ -515,7 +521,9 @@ private void fetchMetrics(JSONObject newJson, List<Map<String, Object>> dataList
 				} else {
 					resultMap.put(entry.getKey(), aggregatedObject.get(APIConstants.FormulaFields.VALUE.getField()));
 				}
-				resultMap.put(groupBy[counter], newJson.get(APIConstants.FormulaFields.KEY.getField()));
+				if(groupBy != null && groupBy.length > 0) {
+					resultMap.put(groupBy[counter], newJson.get(APIConstants.FormulaFields.KEY.getField()));
```
Comment (kumaraswamyS) [inline src/main/java/org/gooru/insights/services/ESDataProcessorServiceImpl.java:525]: @RenukaGurumurthy  groupBy field is passed as null if it has only aggregation so null check is valid,but if it has groupBy then it sure has data so groupBy length > 0 is invalid validation.so remove it.

File: src/main/java/org/gooru/insights/services/BaseAPIServiceImpl.java [hunk 4]
```diff
@@ -483,21 +484,24 @@ public JSONArray formatKeyValueJson(List<Map<String, Object>> dataMap, String ke
...
-		}else if(processedData.get(APIConstants.Hasdatas.HAS_AGGREGATE.check()) || processedData.get(APIConstants.Hasdatas.HAS_GRANULARITY.check())){
+		} else if (processedData.get(APIConstants.Hasdatas.HAS_GRANULARITY.check())){
 			throw new BadRequestException(MessageHandler.getMessage(ErrorConstants.E100, APIConstants.GROUP_BY));
 		}
 		/**
 		 * Range filter validation.Here groupBy field shouldn't be empty
 		 */
 		if(checkNull(requestParamsDTO.getRanges())) {
+			if(!processedData.get(APIConstants.Hasdatas.HAS_GROUPBY.check())){
+				throw new BadRequestException(MessageHandler.getMessage(ErrorConstants.E112, APIConstants.GROUP_BY, APIConstants.RANGE_ATTRIBUTE));
```
Comment (sheebans) [inline src/main/java/org/gooru/insights/services/BaseAPIServiceImpl.java]: Can we use generic reject method  to handle this kind of validation ?

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java [hunk 5]
```diff
@@ -296,23 +299,28 @@ private void groupConcat(List<Map<String,Object>> dataList,List<Map<String,Objec
...
-			filterFields = getBaseAPIService().convertStringtoSet(fields);
-			}else{
-					for(String field : getBaseConnectionService().getDefaultFields().get(indices).split(APIConstants.COMMA)){
-						searchRequestBuilder.addField(field);	
+			if (validatedData.get(Hasdatas.HAS_FEILDS.check())) {
+				if (!requestParamsDTO.getFields().equalsIgnoreCase(APIConstants.WILD_CARD)) {
+					dataKey = EsSources.FIELDS.esSource();
+					String fields = getBusinessLogicService().esFields(indices, requestParamsDTO.getFields());
+					filterFields = getBaseAPIService().convertStringtoSet(fields);
+				} else {
```
Comment (sheebans) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:314]: for (String field : filterFields) {
-                   searchRequestBuilder.addField(field);
-               }

I think above iteration method should be inside  the if condition. // if(!requestParamsDTO.getFields().equalsIgnoreCase(APIConstants.WILD_CARD)){

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java [hunk 6]
```diff
@@ -402,7 +409,12 @@ private void buildBuckets(String traceId,String index, RequestParamsDTO requestP
 		try {
 			TermsBuilder termBuilder = null;
 			String[] groupBy = requestParamsDTO.getGroupBy().split(APIConstants.COMMA);
```
Comment (sheebans) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:411]: by any chance requestParamsDTO.getGroupBy()  is null ?  if so NPE may happen

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java [hunk 7]
```diff
@@ -414,21 +426,16 @@ private void buildBuckets(String traceId,String index, RequestParamsDTO requestP
 					termBuilder = AggregationBuilders.terms(groupBy[i]).field(fieldName);
 				}
 				if (i == groupBy.length - 1) {
-					bucketAggregation(index, requestParamsDTO, termBuilder, metricsName);
+					includeMetricsAggregation(index, requestParamsDTO, filterAggregationBuilder, searchRequestBuilder, termBuilder, null, null, metricsName);
```
Comment (sheebans) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:429]: includeMetricsAggregation(index, requestParamsDTO, filterAggregationBuilder, searchRequestBuilder, termBuilder, null, null, metricsName);

there might be chance of NPE when filterAggregationBuilder is null

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java (hunk 6 above)
Comment (RenukaGurumurthy) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:411]: This method will be called only when requestParamsDTO.getGroupBy() has value

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java (hunk 7 above)
Comment (RenukaGurumurthy) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:429]: We check filterAggregationBuilder inside includeMetricsAggregation method. Based on null check, appropriate aggregationBuilder will be used for further processing.

File: src/main/java/org/gooru/insights/services/BaseESServiceImpl.java (hunk 5 above)
Comment (RenukaGurumurthy) [inline src/main/java/org/gooru/insights/services/BaseESServiceImpl.java:314]: Added changes to assign to filterFields Set

**Pull request merged**